app/src/main/java/org/groundplatform/android/ui/datacollection/DataCollectionHelper.kt

```diff
@@ -0,0 +1,448 @@
+
+    if (validationError != null) {
+      popups.get().ErrorPopup().show(validationError)
+    } else {
+      taskDataHandler.setData(task, taskValue)
+      taskSequenceHandler.invalidateCache()
+      moveToPreviousTask()
+    }
+  }
+
+  fun onNextClicked(taskViewModel: AbstractTaskViewModel) = withReady { st ->
+    validateOrShow(taskViewModel) {
+      val task = taskViewModel.task
+      val value = taskViewModel.taskTaskData.value
+      taskDataHandler.setData(task, value)
+      taskSequenceHandler.invalidateCache()
+
+      if (!taskSequenceHandler.isLastPosition(task.id)) {
+        moveToNextTask()
+      } else {
+        clearDraft()
+        saveChanges(st, getDeltas())
+        _uiState.value = DataCollectionUiState.TaskSubmitted
+      }
+    }
+  }
+
+  fun saveCurrentState() {
+    if (!draftsEnabled) return
+    withReadyOrNull { state ->
+      val taskId = state.currentTaskId
+      getTaskViewModel(taskId)?.let { vm ->
+        taskDataHandler.setData(vm.task, vm.taskTaskData.value)
+        savedStateHandle[TASK_POSITION_ID] = taskId
+        saveDraft(taskId)
+      }
+    }
+  }
+
+  fun clearDraftBlocking() {
+    suppressDrafts()
+    clearDraft()
+  }
+
+  fun moveToPreviousTask() {
+    moveToTask(withReady { taskSequenceHandler.getPreviousTask(it.currentTaskId) })
+  }
+
+  fun moveToNextTask() {
+    moveToTask(withReady { taskSequenceHandler.getNextTask(it.currentTaskId) })
+  }
+
+  fun isFirstPosition(taskId: String): Boolean = withReady {
+    taskSequenceHandler.isFirstPosition(taskId)
+  }
+
+  fun isLastPosition(taskId: String): Boolean = withReady {
+    taskSequenceHandler.isLastPosition(taskId)
+  }
+
+  fun isLastPositionWithValue(task: Task, newValue: TaskData?): Boolean = withReady {
+    if (taskDataHandler.getData(task) == newValue) {
+      taskSequenceHandler.isLastPosition(task.id)
+    } else {
+      taskSequenceHandler.checkIfTaskIsLastWithValue(task.id to newValue)
+    }
+  }
+
+  fun requireSurveyId(): String = withReady { it.surveyId }
+
+  fun isAtFirstTask(): Boolean = withReady { taskSequenceHandler.isFirstPosition(it.currentTaskId) }
+
+  private fun moveToTask(taskId: String) = withReady { st ->
+    val validIds = taskSequenceHandler.getValidTasks().map { it.id }.toSet()
+    val safeId = if (taskId in validIds) taskId else validIds.first()
+
+    savedStateHandle[TASK_POSITION_ID] = safeId
+    saveDraft(safeId)
+
+    val newPos = taskSequenceHandler.getTaskPosition(safeId)
+    _uiState.value = st.copy(currentTaskId = safeId, position = newPos)
+  }
+
+  private fun getDeltas(): List<ValueDelta> {
+    val tasksDataMap = taskDataHandler.dataState.value
+    val validTasks = taskSequenceHandler.getValidTasks()
+    return tasksDataMap
+      .filterKeys { it in validTasks }
+      .map { (task, taskData) -> ValueDelta(task.id, task.type, taskData) }
+  }
+
+  private fun saveChanges(state: DataCollectionUiState.Ready, deltas: List<ValueDelta>) {
```

Comment:
These methods (save changes, move next/previous, submit, etc.) should still remain in the view model class.